jax/linear_util.py

```diff
@@ -208,15 +196,16 @@ def wrap_init(f, params={}):
   return WrappedFun(f, (), (), tuple(sorted(params.items())))
 
 
-def cache(call, max_size=4096):
-  @fastcache.clru_cache(maxsize=max_size)
-  def cached_fun_body(f, args):
-    return call(f, *args), f
-
-  def cached_fun(f, *args):
-    ans, f_prev = cached_fun_body(f, args)
-    if id(f_prev) != id(f):
-      f.populate_stores(f_prev)
+def cache(call):
+  caches = weakref.WeakKeyDictionary()
+  def memoized_fun(f, *args):
+    cache = caches.setdefault(f.f, {})
+    key = (f.transforms, f.params, args)
+    if key in cache:
+      ans, stores = cache[key]
+      f.populate_stores(stores)
+    else:
+      ans = call(f, *args)
+      cache[key] = (ans, f.stores)
```

Comment:
I spent a little while thinking about possible race conditions with multi-threading and convinced myself that this is fine. We don't do any clean-up, so `cache[key]` is guaranteed to be valid if `key in cache`. In the worst case, code gets compiled twice.

Fun fact: `dict.setdefault` is atomic, but didn't always used to be: https://bugs.python.org/issue13521

Reply:
I'm glad you thought about this hard case, because I didn't :)